README.md
# Homo Sapiens Time

A simple time utility for Node.js

Human-style time descriptions to milliseconds and vice versa.

_Quite a Simple One_

## Installation
```bash
npm install homo-sapiens-time
#or
yarn add homo-sapiens-time
```

## Usage
#### Node
```js
const hst = require("homo-sapiens-time")
```
#### HTML/JS
```html
<script type="text/javascript" src="homo-sapiens-time.js"> </script>
<div id="message"> </div>
<script> 
    var x = msToTimeString(31557601000, {auto: false, units: ['seconds', 'year'], sortUnits: true}); 
    document.getElementById("message").innerHTML=x;
</script>
```

## Description
There are three main functions.

- `msToTimeString`
- `timeStringToMs`
- `impreciseDurationAddedToNow`

Defined time units are not as precise as calendar. 
| Unit          | Milliseconds | Equivalent To |
|---------------|--------------|---------------|
| 1 year        | 31557600000  | 365.25 days   |     
| 1 month       | 2629800000   | 30.4375 days  |      
| 1 week        | 604800000    | 7 days        |     
| 1 day         | 86400000     | 24 hours      |     
| 1 hour        | 3600000      | 60 minutes    |    
| 1 minute      | 60000        | 60 seconds    |     
| 1 seconds     | 1000         | 1000 ms       |     
| 1 millisecond | 1            | 1 ms          |  

## Time String Format
A single **time part** consists of the following format:
- Positive or Negative Integer
- Time Unit

There can only be one or zero spaces between each part.
**Time part** can repeat as much as needed.

All available time units are: 

| Plural        | Singular     | Short | Maps To|
|---------------|--------------|-------|:------:|
| years         | year         | y     | 'y'    |     
| months        | month        | mo    | 'mo'   |      
| weeks         | week         | w     | 'w'    |     
| days          | day          | d     | 'd'    |     
| hours         | hour         | h     | 'h'    |    
| minutes       | minute       | m     | 'm'    |     
| seconds       | second       | s     | 's'    |     
| milliseconds  | millisecond  | ms    | 'ms'   |     

Plural, singular or short version of unit doesn't affect calculation,
it just affects printed text. Integer value is not checked for plural or singular units either.

### Example

- "1year" = "1 year" = "1 y"
- "2 year 3 months"
- "31557601000 milliseconds"
- "1year -6 month" = "6 months"

-----------------------

## 1. msToTimeString

### Summary
This function takes `ms` and converts it to string that consists of given units.

This method has FIFO execution for `units` which means order of units matter.
For example, if `ms` is given as the first value of `units` array then result will be same as `ms` parameter's value since `ms` in `units` will consume all value.

This function will detect any unit type typos given and will suggestions. For example:
`msToTimeString(31557601000, {auto: false, units: ['seconds', 'yeer']})` will produce
```text
Error! Given unit `yeer` is not valid.
           Did you mean `year`?
```
String similarity function is taken from [David from StackOverflow](https://stackoverflow.com/a/36566052).
Thanks David. 👍

### Parameters

##### 1.1 `ms`
Number containing time in milliseconds.

##### 1.2 `opts`
```JS
opts = {
    auto: boolean,
    units: array of string,
    showEmpty: boolean,
    sortUnits: boolean
}
```
##### 1.2.1 `opts.auto` 

when set to `true` units will be overwritten as `['year','month','week','day','hour','minute','seconds','ms']`

##### 1.2.2 `opts.units`
any valid set of time units. Following are the available units:
```JS
const units = {
    years : 'y',
    year : 'y',
    y : 'y',

    months : 'mo',
    month : 'mo',
    mo : 'mo',

    weeks: 'w',
    week: 'w',
    w: 'w',

    days: 'd',
    day: 'd',
    d: 'd',

    hours: 'h',
    hour: 'h',
    h: 'h',

    minutes: 'm',
    minute: 'm',
    m: 'm',

    seconds : 's',
    second : 's',
    s : 's',

    milliseconds : 'ms',
    millisecond : 'ms',
    ms : 'ms'
}
```
> Default value is `['year','month','week','day','hour','minute','second','ms']` if `auto` is `true` else `['hour','minute','second']`.

##### 1.2.3 `opts.showEmpty`
when set to `true`, if a unit with no value exists in the result, it will be written regardless. 

> Default value is `false`.

##### 1.2.4 `opts.sortUnits`
Since order of `opts.units` matters, you can sort units to make longer time units (like year being longer than month) to have higher priority.
Priority: `y > mo > w > d > h > m > s > ms`
Units with higher priority will be closer to beginning of array.

> Default value is `false`.

##### Return
Returns time string in **Time String Format** or `null` if any error occurred.

#### Examples
```JS
msToTimeString(31557601000, {auto: false, units: ['seconds', 'year'], sortUnits: true}) == '1 year 1 seconds'
msToTimeString(31557601001, {auto : true, units: [ 'year', 'month' ], showEmpty: false}) == '1 year 1 seconds 1 ms'
msToTimeString(74580249002, {auto: true}) == '2 year 4 month 1 week 3 day 22 hour 44 minute 9 second 2 ms'

```

## 2. timeStringToMs
### Summary
Takes a time string similar to ones produced in `msToTimeString` and converts it to milliseconds.

### Parameters

##### 1. `str`:
String containing time in **Time String Format**.

##### Return
Returns ms converted from `str`.

#### Examples
```JS
timeStringToMs("1 year -6 month") == timeStringToMs("6 month")
timeStringToMs("2 years 4 months 10 days 22 hours 44 minutes 9 seconds 2 ms") == 74580249002
timeStringToMs("1 year 1 month 1 week 1 day 1 hour -1 hour -1 day -1mo -1y") == timeStringToMs("1 week")
timeStringToMs("1 week") == timeStringToMs("7 days")
```

## 3. impreciseDurationAddedToNow

### Summary

This method adds given time string `str` to `Date.now()` value.
Time string must be in **Time String Format**.

Internally calls `timeStringToMs` to convert `str` to ms.

### Parameters

##### 1. `str`:
String containing time in **Time String Format**.


##### Return
Millisecond value of `Date.now() + timeStringToMs(str)`.

## Notes
> This library is perfect for durations. Since it is not precise I highly **recommend being careful** about using
> it for precise calendar operations.


CHANGELOG.md
# Changelog

## [0.6.1] - 2022-02-24
- Fixed second being seconds, now it is second :)
- Added more documentation and examples to README.md.

## [0.5.0] - 2022-02-23
### Added
- Very first public release.
- Added timeStringToMs function that converts time string to milliseconds.
- Added msToTimeString function that converts milliseconds to time string.
- Added impreciseDurationAddedToNow function that adds time string to Date.now()


test.js
// import {msToTimeString} from "./homo-sapiens-time.js"
const lib = require("./homo-sapiens-time.js")

/**
 * Following 8 tests will be run for same value to test options.
 * options = { true , obj , true  }
 * options = { true , obj , false }
 * options = { true , none, true  }
 * options = { true , none, false }
 * options = { false, obj , true  }
 * options = { false, obj , false }
 * options = { false, none, true  }
 * options = { false, none, false }
 */

test('Converts 31557601001 to 1 year 0 month 0 week 0 day 0 hour 0 minute 1 second 1 ms', () => {
    var a = lib.msToTimeString(31557601001, {auto : true, units: [ 'year', 'month' ], showEmpty: true});
    console.log(a);

    expect(a).toBe('1 year 0 month 0 week 0 day 0 hour 0 minute 1 second 1 ms');
});

test('Converts 31557601001 to 1 year 1 second 1 ms', () => {
    var a = lib.msToTimeString(31557601001, {auto : true, units: [ 'year', 'month' ], showEmpty: false});
    console.log(a);

    expect(a).toBe('1 year 1 second 1 ms');
});

test('Converts 31557601001 to 1 year 0 month 0 week 0 day 0 hour 0 minute 1 second 1 ms', () => {
    var a = lib.msToTimeString(31557601001, {auto : true, showEmpty: true});
    console.log(a);

    expect(a).toBe('1 year 0 month 0 week 0 day 0 hour 0 minute 1 second 1 ms');
});

test('Converts 31557601001 to 1 year 1 second 1 ms', () => {
    var a = lib.msToTimeString(31557601001, {auto : true, showEmpty: false});
    console.log(a);

    expect(a).toBe('1 year 1 second 1 ms');
});

test('Converts 31557601001 to 1 year 0 month', () => {
    var a = lib.msToTimeString(31557601001, {auto : false, units: [ 'year', 'month' ], showEmpty: true});
    console.log(a);

    expect(a).toBe('1 year 0 month');
});

test('Converts 31557601001 to 1 year', () => {
    var a = lib.msToTimeString(31557601001, {auto : false, units: [ 'year', 'month' ], showEmpty: false});
    console.log(a);

    expect(a).toBe('1 year');
});

test('Converts 31557601001 to 8766 hour 0 minute 1 second', () => {
    var a = lib.msToTimeString(31557601001, {auto : false, showEmpty: true});
    console.log(a);

    expect(a).toBe('8766 hour 0 minute 1 second');
});

test('Converts 31557601001 to 8766 hour 1 second', () => {
    var a = lib.msToTimeString(31557601001, {auto : false, showEmpty: false});
    console.log(a);

    expect(a).toBe('8766 hour 1 second');
});


test('Converts 1645307362744 to 8766 hour 1 second', () => {
    var a = lib.msToTimeString(1645307362744, {auto : false, units: ['year','month','day','hour','minute','seconds','ms'], showEmpty: false});
    console.log(a);

    expect(a).toBe('52 year 1 month 19 day 11 hour 19 minute 22 seconds 744 ms');
});

test('Add time to now', () => {
    // 1 year 1 month = 34187400000
    var now = Date.now();
    var a = lib.impreciseDurationAddedToNow("1 year 1 month");
    console.log(a);

    let diff = a - now;
    console.log(diff);

    //            -2 seconds                    1 seconds
    var result  = 34187398000 < diff && diff <= 34187401000;
    expect(result).toBe(true);

});

test('Time added to now converted to string', () => {
    // 1 year 1 month = 34187400000
    let time_str = "1 year 1 month";
    var now = Date.now();
    var a = lib.impreciseDurationAddedToNow(time_str);
    console.log(a);

    let diff = a - now;
    console.log(diff);
    //            -2 seconds                    1 seconds
    var result  = 34187398000 < diff && diff <= 34187401000;
    expect(result).toBe(true);
    var str = lib.msToTimeString(diff, {auto: false, showEmpty: true, units: ['year', 'month']});
    expect(str).toBe(time_str);
});


test('Add negative time to now', () => {
    // -1 minute = -60000
    var now = Date.now();
    var a = lib.impreciseDurationAddedToNow("-1 minute");
    console.log(a);
    let diff = now - a;
    console.log(diff);
    var result  = 0 < diff && diff <= 60000;
    expect(result).toBe(true);
});


test('1 year - 6 month = 6 month', () => {
    var a = lib.timeStringToMs("1 year -6 month");
    var b = lib.timeStringToMs("6 month");

    console.log(`a-> ${a} === ${b} <-b`)

    expect(a).toBe(b);
});

test('Wrong unit name in msToTimeString', () => {
    var a = lib.msToTimeString(31557601000, {auto: false, units: ['seconds', 'yeer']});

    expect(a).toBe(null);
});

test('Sort units', () => {
    var a = lib.msToTimeString(31557601000, {auto: false, units: ['seconds', 'year']});
    expect(a).toBe('31557601 seconds');

    var b = lib.msToTimeString(31557601000, {auto: false, units: ['seconds', 'year'], sortUnits: true});
    expect(b).toBe('1 year 1 seconds');
});

test('2 year 2 seconds', () => {
    var b = lib.msToTimeString(63115202000, {auto: false, units: ['seconds', 'year'], sortUnits: true});
    expect(b).toBe('2 year 2 seconds');
});

test('74580249002 ms to auto', () => {
    var a = lib.msToTimeString(74580249002, {auto: true});
    expect(a).toBe('2 year 4 month 1 week 3 day 22 hour 44 minute 9 second 2 ms');
})

test('74580249002 ms to 2 year 4 month 10 day 22 hour 44 minute 9 second 2 ms', () => {
    var a = lib.msToTimeString(74580249002, {auto: false, units: [ 'year', 'month', 'day', 'hour', 'minute', 'second', 'ms' ]});
    expect(a).toBe('2 year 4 month 10 day 22 hour 44 minute 9 second 2 ms');
})

test('2 years 4 months 10 days 22 hours 44 minutes 9 seconds 2 ms to 74580249002', () => {
    var a = lib.timeStringToMs("2 years 4 months 10 days 22 hours 44 minutes 9 seconds 2 ms");
    expect(a).toBe(74580249002);
});

test('1 year 1 month 1 week 1 day 1 hour -1 hour -1 day -1mo -1y == 1 week', () => {
    var a = lib.timeStringToMs("1 year 1 month 1 week 1 day 1 hour -1 hour -1 day -1mo -1y");
    var b = lib.timeStringToMs("1 week");
    var c = lib.timeStringToMs("7 days");
    expect(a).toBe(b);
    expect(b).toBe(c);
});

test('Date Now', () => {
    var a = lib.msToTimeString(Date.now(), {auto: false, units: ['year', 'month', 'day', 'hour', 'minute', 'second', 'ms']});
    console.log(a)
});


homo-sapiens-time.js
// {
//     [years, months, weeks, days, hours, minutes, seconds, milliseconds]
//     [year, month, week, day, hour, minute, second, millisecond]   
//     [y, mo, w, d, h, m, s, ms]
// }

// milliseconds|seconds|minutes|hours|days|weeks|months|years|millisecond|second|minute|hour|day|week|month|year|ms|mo|m|s|h|d|y|w

const timeConstants = {
    y : 31557600000,
    mo: 2629800000,
    w : 604800000,
    d : 86400000,
    h : 3600000,
    m : 60000,
    s : 1000,
    ms: 1
}

const units = {
    years : 'y',
    year : 'y',
    y : 'y',

    months : 'mo',
    month : 'mo',
    mo : 'mo',

    weeks: 'w',
    week: 'w',
    w: 'w',

    days: 'd',
    day: 'd',
    d: 'd',

    hours: 'h',
    hour: 'h',
    h: 'h',

    minutes: 'm',
    minute: 'm',
    m: 'm',

    seconds : 's',
    second : 's',
    s : 's',

    milliseconds : 'ms',
    millisecond : 'ms',
    ms : 'ms'
}

const msConversionRegex = new RegExp('((-?\\d+)\\s*(milliseconds|seconds|minutes|hours|days|weeks|months|years|millisecond|second|minute|hour|day|week|month|year|ms|mo|m|s|h|d|y|w)\\s*)', 'g')

// https://stackoverflow.com/questions/10473745/compare-strings-javascript-return-of-likely
function similarity(s1, s2) {
    var longer = s1;
    var shorter = s2;
    if (s1.length < s2.length) {
      longer = s2;
      shorter = s1;
    }
    var longerLength = longer.length;
    if (longerLength == 0) {
      return 1.0;
    }
    return (longerLength - editDistance(longer, shorter)) / parseFloat(longerLength);
}

// https://stackoverflow.com/questions/10473745/compare-strings-javascript-return-of-likely
function editDistance(s1, s2) {
    s1 = s1.toLowerCase();
    s2 = s2.toLowerCase();
  
    var costs = new Array();
    for (var i = 0; i <= s1.length; i++) {
      var lastValue = i;
      for (var j = 0; j <= s2.length; j++) {
        if (i == 0)
          costs[j] = j;
        else {
          if (j > 0) {
            var newValue = costs[j - 1];
            if (s1.charAt(i - 1) != s2.charAt(j - 1))
              newValue = Math.min(Math.min(newValue, lastValue),
                costs[j]) + 1;
            costs[j - 1] = lastValue;
            lastValue = newValue;
          }
        }
      }
      if (i > 0)
        costs[s2.length] = lastValue;
    }
    return costs[s2.length];
}

const getMSEquivalent = (type, count) => {
    return count * timeConstants[units[type]]
}

function timeStringToMs(str) {
    let milliseconds = 0;
    let myArray;
    while ((myArray = msConversionRegex.exec(str)) !== null) {
        var type = myArray[3];
        var count = parseInt(myArray[2]);
        milliseconds += getMSEquivalent(type, count);
    }
    return milliseconds;
}

/**
 * Takes `ms` and converts it to string that consists of given units.
 * @param {number}  ms 
 * @param {boolean} [opts.auto=true] automatically converts time with 'year','month','week','day','hour','minute','seconds','ms' 
 * @param {array}   [opts.units=['h', 'm' ,'s']] array of wanted units
 * @param {boolean} [opts.showEmpty=false] show even if given unit is empty
 * @param {boolean} [opts.sortUnits=false] sorts units for execution priority. 
 * @return null if given units are wrong. (prints error to console)
 * 
 * Unlike https://github.com/shime/humanize-time/blob/master/index.js
 * 
 * Unlike https://github.com/EvanHahn/HumanizeDuration.js/blob/main/humanize-duration.js
 */
function msToTimeString(ms, opts) {
    if (!opts) {
        opts = {};
    }

    var options = {
        auto      : opts.auto      ?? true,
        // Array assignment creates reference not copy
        // Why spread first: options.units and opts.units are pointing to same object
        units     : (opts.auto ? ['year','month','week','day','hour','minute','second','ms'] : ( opts.units ? [...opts.units] : ['hour','minute','second'])),   
        showEmpty : opts.showEmpty ?? false,
        sortUnits : opts.sortUnits ?? false
    }

    if (options.sortUnits) {
        options.units.sort((firstEl, secondEl) => {
            if( Object.keys(units).indexOf(firstEl) < Object.keys(units).indexOf(secondEl) ) {
                return -1;
            }
            
            if( Object.keys(units).indexOf(firstEl) > Object.keys(units).indexOf(secondEl) ) {
                return 1;
            }

            return 0;
        })
    }

    // console.log("IMPORTANT: " + JSON.stringify(options.units));

    var time  = {
        y  : 0,
        mo : 0,
        w  : 0,
        d  : 0,
        h  : 0,
        m  : 0,
        s  : 0,
        ms : 0
    }

    const defaultUnits = [ 'y', 'mo' , 'w' , 'd', 'h', 'm' ,'s', 'ms' ];
    // var currentUnits = options.auto ? defaultUnits : options.units;   
    var currentUnits = options.units;   

    let time_str = '';
    let runnable = true;

    currentUnits.forEach(currentUnit => {
        // Verify options
        if (defaultUnits.indexOf(units[currentUnit]) < 0 ) {
            let max  = 0;
            let unit = 0;

            Object.keys(units).forEach(x => {
                let currentValue = similarity(currentUnit, x);

                if (currentValue >= max) {
                    max = currentValue;
                    unit = x;
                }
            });

            console.error(`\nError! Given unit \`${currentUnit}\` is not valid.\n       Did you mean \`${unit}\`?`);
            // process.exit(1);
            runnable = false;
        }
    });

    if (!runnable)
        return null;

    currentUnits.forEach(currentUnit => {
        // Run units.
        while ( ms >= timeConstants[units[currentUnit]] ) {
            ms -= timeConstants[units[currentUnit]];
            time[units[currentUnit]] += 1;
        }

        if (options.showEmpty) {
            time_str += time[units[currentUnit]] + ' ' + currentUnit + ' ';
        } else if (time[units[currentUnit]] > 0) {
            time_str += time[units[currentUnit]] + ' ' + currentUnit + ' ';
        }
        
    });
    // console.log(`Time: ${JSON.stringify(time)} | Current Units: ${JSON.stringify(currentUnits)}`);        
    // console.log(time_str);
    
    return time_str.trim();

}

function impreciseDurationAddedToNow(str) {
    let time = timeStringToMs(str);
    let now = Date.now();
    let added = time + now;
    return added;
}

module.exports = {
    timeStringToMs: timeStringToMs,
    msToTimeString: msToTimeString,
    impreciseDurationAddedToNow : impreciseDurationAddedToNow
};